src/App.tsx

```diff
@@ -1,8 +1,53 @@
-import React from 'react';
+import React, { Component } from 'react';
 import './App.css';
+import { getGoods } from './api/getGoods';
+import { Good } from './interfaces';
+import { GoodsList } from './components/GoodsList';
 
-const App = () => (
-  <h1>Dynamic list of Goods</h1>
-);
+interface State {
+  goods: Good[];
+}
+
+class App extends Component<{}, State> {
+  state: State = {
+    goods: [],
+  };
+
+  handleClickAllGoods = () => {
+    getGoods().then(goods => this.setState({
+      goods,
+    }));
+  };
+
+  handleClickFirstFive = () => {
+    getGoods().then(goods => this.setState({
+      goods: goods
+        .sort((a, b) => a.name.localeCompare(b.name))
+        .splice(0, 5),
+    }));
+  };
+
+  handleClickOnlyRed = () => {
+    getGoods().then(goods => this.setState({
+      goods: goods.filter((good) => good.color === 'red'),
+    }));
+  };
+
+  render() {
+    const { goods } = this.state;
+
+    return (
+      <div className="container is-fluid">
+        <h1 className="title">List of goods</h1>
+        <GoodsList
+          goods={goods}
+          handleClickAllGoods={this.handleClickAllGoods}
+          handleClickFirstFive={this.handleClickFirstFive}
+          handleClickOnlyRed={this.handleClickOnlyRed}
+        />
```

Comment:
Single responsibility principle - GoodsList - should be responsible for 1 thing - render the list
create a separate component for buttons or render them here

Also, you do not need state in App if buttons and list are in GoodsList